crates/contracts/src/world/world_contract.cairo

```diff
@@ -1043,6 +1050,43 @@ pub mod world {
             }
         }
 
+        /// Panics if a resource is not upgradable.
+        ///
+        /// Upgradable means:
+        /// - the layout type must remain the same (Struct or Fixed),
+        /// - existing fields cannot be changed or moved inside the resource,
+        /// - new fields can only be appended at the end of the resource.
+        ///
+        /// # Arguments
+        ///   * `namespace` - the namespace of the resource.
+        ///   * `name` - the name of the resource.
+        ///   * `prev_address` - the address of the current resource.
+        ///   * `new_address` - the address of the newly deployed resource.
+        ///
+        fn assert_resource_upgradability(
+            self: @ContractState,
+            namespace: @ByteArray,
+            name: @ByteArray,
+            prev_address: ContractAddress,
+            new_address: ContractAddress
+        ) {
+            let resource = IStoredResourceDispatcher { contract_address: prev_address };
+            let old_layout = resource.layout();
+            let old_schema = resource.schema();
+
+            let new_resource = IStoredResourceDispatcher { contract_address: new_address };
+            let new_layout = new_resource.layout();
+            let new_schema = new_resource.schema();
+
+            if !new_layout.is_same_type_of(@old_layout) {
+                panic_with_byte_array(@errors::invalid_resource_layout_upgrade(namespace, name));
+            }
+
+            if !new_schema.is_an_upgrade_of(@old_schema) {
+                panic_with_byte_array(@errors::invalid_resource_schema_upgrade(namespace, name));
+            }
```

Comment:
New function without side effect only to check if a model/event is a good candidate for upgradeability.